feat(component): added Button component to control Menu

souporserious opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: http://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->
**What**:
A `Button` component that applies proper ARIA attributes and allows controlling the opening and closing of the menu as well as keyboard navigation of the list of items.
<!-- Why are these changes necessary? -->
**Why**:
This helps us build accessible dropdown + autocomplete solutions.
<!-- How were these changes implemented? -->
**How**:
I added a `Button` component that mimics some of the behavior of the `Input`.
<!-- Have you done all of these things?  -->
**Checklist**:
<!-- add "N/A" to the end of each line that's irrelevant to your changes -->
<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->
- [x] Documentation
- [ ] Tests
- [x] Ready to be merged <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->
- [ ] Added myself to contributors table <!-- this is optional, see the contributing guidelines for instructions -->

<!-- feel free to add additional comments -->
Another addition to the API 😇 I think this will help with having an accessible button handy to control the `Menu`. I used it real quick in your semantic-ui demo and it worked nice 👌 applies all the proper ARIA attributes and allows for keyboard navigation.

kentcdodds commented on examples/pages/semantic-ui/index.js:
```diff
@@ -148,26 +148,18 @@ function SemanticUIAutocomplete() {
-                onClick={clearSelection}
-                aria-label="clear selection"
-                >
-                <XIcon />
```
Looks like we lost this functionality here. Can we keep the `XIcon` somehow?

souporserious replied:
Oops! Yep, I was just testing it out real quick. I'll add it back.

kentcdodds commented on src/button.js:
```diff
@@ -0,0 +1,92 @@
+  }
+
+  keyDownHandlers = {
+    ArrowDown(event) {
```
Oh nice! I was thinking about this! I was actually thinking that once the dropdown is expanded we should focus on the input instead. What do you think about that?

souporserious replied:
Yes :) I had that working for the Label component and then I started making this one 😆 I'll add it here.

kentcdodds replied:
If we make clicking this button automatically focus the input, then we won't need these event handlers right? Should we do that? Or is that too opinionated?

kentcdodds commented on src/button.js:
```diff
@@ -0,0 +1,92 @@
+    this.autocomplete.toggleMenu()
+  }
+
+  focusButton = () => {
```
Should we expose this method in the `Controller`?

souporserious replied:
Yup 👍

souporserious commented:
Alright, I think this one is good to go now :) I removed the `focusButton` method because I don't think it will be necessary and would like to keep the API surface small if we can.

I also added a returnFocus method to restore the last focused item when the menu closes, and made sure the status menu is created and destroyed when the menu opens/closes.

kentcdodds commented on README.md:
```diff
@@ -120,6 +120,10 @@ This is called with an object with the properties listed below:
 | `selectItemAtIndex`     | `function(index: number)`  | selects the item at the given index                                                                              |
 | `selectHighlightedItem` | `function()`               | selects the item that is currently highlighted                                                                   |
 
+### Autocomplete.Button
```
We should document the props as we do for the other components

souporserious replied:
There aren't any props 😁 we pull `disabled` off, but I think it's fine since that's a normal prop for button.

kentcdodds replied:
Ah, gotcha :+1: good call :)

kentcdodds commented on examples/pages/semantic-ui/index.js:
```diff
@@ -160,7 +160,6 @@ function SemanticUIAutocomplete() {
               </ControllerButton>
               : <ControllerButton
-                onClick={toggleMenu}
                 aria-label={isOpen ? 'close menu' : 'open menu'}
```
Can we remove this? The built in bottom should now handle all a11y for us right?

souporserious replied:
Done 👍

kentcdodds commented on src/input.js:
```diff
@@ -105,6 +107,7 @@ class Input extends Component {
         aria-expanded={isOpen}
         autoComplete="off"
         value={(inputValue === null ? selectedItemValue : inputValue) || ''}
+        autoFocus={this.menu}
```
Interesting! What does this do? Should probably pass this to the boolean constructor to make this more explicit, but what's it for?

souporserious replied:
It's just checking for the presence of `Menu` and then evaluates `autoFocus` to true. So no matter if `Button` opens it or not, it will autoFocus based on being a child of `Menu`.

kentcdodds replied:
Gotcha. Do you mind wrapping it in `Boolean(this.menu)`? I just prefer to avoid type coercion in certain scenarios (this being one of them) :smile:

kentcdodds commented on src/menu.js:
```diff
@@ -103,31 +103,50 @@ class Menu extends Component {
-      />,
-      this._statusNode,
-    )
+    this._statusNode &&
```
I'd prefer an if statement here 👍

kentcdodds commented on src/menu.js:
```diff
@@ -103,31 +103,50 @@ class Menu extends Component {
+    const lastActiveElement = document.activeElement
+    this.returnFocus = () => {
+      // only return focus if a non-focusable item was selected when closing
+      if (document.activeElement === document.body) {
```
Hmmm, how does this work? Will this ever be true?

souporserious replied:
Yes as long as you don't click another focusable item this will return true. So imagine you click another menu button, you would want that to be focused and not return focus to the button that opened it. I hope that makes sense.

kentcdodds replied:
I see, so you're expecting that if they don't click on a focusable element, the body will receive focus (which is correct IIUC) :+1: Thanks!

souporserious commented:
Alright, now I _think_ it's all good to go :) pertaining to your question above (I couldn't reply for some reason), having the key handlers is beneficial to allow selecting elements from the list even when the input is not present. This is where I was saying it could be used for more than just an autocomplete 😇

kentcdodds commented on src/button.js:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  keyDownHandlers = {
+    ArrowDown(event) {
```
What did you think of this?

> If we make clicking this button automatically focus the input, then we won't need these event handlers right? Should we do that? Or is that too opinionated?

souporserious replied:
I think it's beneficial since it allows you to interact with the list of items. Similar to how react-aria-menubutton does when it receives focus.

kentcdodds replied:
Sounds good 👍

kentcdodds commented on src/button.js:
```diff
@@ -0,0 +1,87 @@
+        {...rest}
+        aria-expanded={isOpen}
+        aria-haspopup={true}
+        disabled={disabled}
```
If all we're doing is just forwarding this do we even need to specify it here? This would get applied with `{...this.props}` if we just did that instead.

souporserious replied:
Yup, totally. I realized I was using disabled for tabIndex, but we don't need that since it's a real button so I removed it which means we can just forward all props 🎉

kentcdodds commented on src/button.js:
```diff
@@ -0,0 +1,87 @@
+      <button
+        role="button"
+        aria-label={isOpen ? 'close menu' : 'open menu'}
+        onClick={this.handleClick}
```
This should go below the spread of props so it doesn't get overridden. We compose this handler so users will be able to specify a custom `onClick`.

souporserious replied:
This was on purpose. Things like `clearSelection` won't work properly. Should we document that?

kentcdodds commented on src/button.js:
```diff
@@ -0,0 +1,87 @@
+        onClick={this.handleClick}
+        {...rest}
+        aria-expanded={isOpen}
+        aria-haspopup={true}
```
These `aria-` props should probably go above the spread.

souporserious replied:
Wasn't sure if they should be able to be overridden or not. I'll put them back 👍

kentcdodds commented on src/input.js:
```diff
@@ -21,6 +22,7 @@ class Input extends Component {
   constructor(props, context) {
     super(props, context)
     this.autocomplete = this.context[AUTOCOMPLETE_CONTEXT]
+    this.autoFocus = typeof this.context[MENU_CONTEXT] !== 'undefined'
```
This changes the behavior from before... Now it will almost _always_ be `false` because the menu is often constructed _after_ the `input`. Could we just do `autoFocus={Boolean(this.menu)}`?

souporserious replied:
I _think_ this should be fine since this will only evaluate true if it's a child of `Menu` which means `Menu` needs to render first right?

kentcdodds replied:
Ah! Gotcha! You're right!

kentcdodds commented:
```
~/Developer/react-autocompletely (master)
🐑  $ npm publish --tag beta
+ react-autocompletely@1.0.0-beta.2
```

souporserious commented:
🙌 🎉  thanks so much! Going to start making examples now 😎

kentcdodds commented:
Thank YOU!
